### PR TITLE
server: set multiple concurrentReadTx instances share one txReadBuffer. Approach #2

### PR DESCRIPTION
**Pull request by wilsonwang371**
This is a different approach for the similar issue trying to resolve in #12933

Currently, the concurrentReadTx does a copy of the txReadBuffer inside backend readTx. However, this buffer is never modified by concurrentReadTx. Therefore, we can just use a single copy of the txReadBuffer as long as there are no new changes committed to backend DB.

Therefore, we use a cached txReadBuffer to avoid excessive deep copy operations for the new read transaction.

### CODE REVIEW

**Comment by wilsonwang371** [comment]
This is a patch inspired by #12529

**Comment by wilsonwang371** [comment]
I am going to run a performance evaluation of this patch later.

**Comment by wilsonwang371** [comment]
@gyuho @ptabor @jingyih

Need u guys to take a look at this patch too. This patch can significantly increase the read qps performance.

File: server/mvcc/backend/tx_buffer.go [hunk 1]
```diff
@@ -88,12 +93,19 @@ func (txr *txReadBuffer) ForEach(bucketName []byte, visitor func(k, v []byte) er
 	return nil
 }
 
+// generate a new cached txReadBuffer, caller should acquire backend.readTx.RLock()
```
Comment (ptabor) [inline server/mvcc/backend/tx_buffer.go:96]: "caller should acquire" -> by convention we call such methods "unsafeGenerateNew..."

Comment (wilsonwang371) [reply server/mvcc/backend/tx_buffer.go:96]: done

**Comment by wilsonwang371** [comment]
Just realized this patch is shorter than #12933, but the performance will be most probably lower than #12933.

The reason for this is that write requests will have to do a copy of the readTx txReadBuffer each time doing a commit.

But I will still upload a performance evaluation for this patch too.

**Comment by wilsonwang371** [comment]
Never mind about this patch. Its initial performance is horrible.

**Pull request closed without merging**